# **Result Summary and Utility Functions**

In [1]:
from pyXLMS import __version__

print(f"Installed pyXLMS version: {__version__}")

Installed pyXLMS version: 1.5.1


In [2]:
from pyXLMS import parser
from pyXLMS import transform

All data transformation functionality is available via the `transform` submodule. We also import the `parser` submodule here for reading result files.

In [3]:
parser_result = parser.read(
    "../../data/ms_annika/XLpeplib_Beveridge_QEx-HFX_DSS_R1.pdResult",
    engine="MS Annika",
    crosslinker="DSS",
)

Reading MS Annika crosslinks...: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [00:00<00:00, 13044.83it/s]


We read crosslink-spectrum-matches and crosslinks using the [generic parser](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.parser.html#pyXLMS.parser.read) from a single `.pdResult` file.

In [4]:
csms = parser_result["crosslink-spectrum-matches"]
xls = parser_result["crosslinks"]

For easier access we assign our crosslink-spectrum-matches to the variable `csms` and our crosslinks to the variable `xls`.

## **Result Summary**

In [5]:
csms_summary = transform.summary(csms)

Number of CSMs: 826.0
Number of unique CSMs: 826.0
Number of intra CSMs: 803.0
Number of inter CSMs: 23.0
Number of target-target CSMs: 786.0
Number of target-decoy CSMs: 39.0
Number of decoy-decoy CSMs: 1.0
Minimum CSM score: 1.1132827525593785
Maximum CSM score: 452.9861536355926


We can get a summary of our crosslink-spectrum-matches by calling `transform.summary()` and passing the crosslink-spectrum-matches as the first argument. The function returns a dictionary containing summary statistics. You can read more about the `summary()` function and all its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.summary.summary).

In [6]:
csms_summary

{'Number of CSMs': 826.0,
 'Number of unique CSMs': 826.0,
 'Number of intra CSMs': 803.0,
 'Number of inter CSMs': 23.0,
 'Number of target-target CSMs': 786.0,
 'Number of target-decoy CSMs': 39.0,
 'Number of decoy-decoy CSMs': 1.0,
 'Minimum CSM score': 1.1132827525593785,
 'Maximum CSM score': 452.9861536355926}

The dictionary returned by `summary()`. By default all summary statistics are also printed to `stdout`.

In [7]:
validated_csms = transform.validate(csms)
_ = transform.summary(validated_csms)

Iterating over scores for FDR calculation...:  15%|███████████▊                                                                     | 121/826 [00:00<00:00, 30246.78it/s]

Number of CSMs: 705.0
Number of unique CSMs: 705.0
Number of intra CSMs: 701.0
Number of inter CSMs: 4.0
Number of target-target CSMs: 699.0
Number of target-decoy CSMs: 6.0
Number of decoy-decoy CSMs: 0.0
Minimum CSM score: 34.188549584398956
Maximum CSM score: 452.9861536355926


You can also repeatedly call `summary()` after data transformation steps to see how your results change, here as an example demonstrated after validation of crosslink-spectrum-matches.

In [8]:
_ = transform.summary(xls)

Number of crosslinks: 300.0
Number of unique crosslinks by peptide: 300.0
Number of unique crosslinks by protein: 298.0
Number of intra crosslinks: 279.0
Number of inter crosslinks: 21.0
Number of target-target crosslinks: 265.0
Number of target-decoy crosslinks: 0.0
Number of decoy-decoy crosslinks: 35.0
Minimum crosslink score: 1.1132827525593785
Maximum crosslink score: 452.9861536355926


Similarly, we can get a summary of our crosslinks by calling `transform.summary()` and passing the crosslinks as the first argument.

In [9]:
_ = transform.summary(parser_result)

Number of CSMs: 826.0
Number of unique CSMs: 826.0
Number of intra CSMs: 803.0
Number of inter CSMs: 23.0
Number of target-target CSMs: 786.0
Number of target-decoy CSMs: 39.0
Number of decoy-decoy CSMs: 1.0
Minimum CSM score: 1.1132827525593785
Maximum CSM score: 452.9861536355926
Number of crosslinks: 300.0
Number of unique crosslinks by peptide: 300.0
Number of unique crosslinks by protein: 298.0
Number of intra crosslinks: 279.0
Number of inter crosslinks: 21.0
Number of target-target crosslinks: 265.0
Number of target-decoy crosslinks: 0.0
Number of decoy-decoy crosslinks: 35.0
Minimum crosslink score: 1.1132827525593785
Maximum crosslink score: 452.9861536355926


Lastly, we can also get a summary of our `parser_result` by calling `transform.summary()` and passing the `parser_result` as the first argument.

## **Crosslink-Spectrum-Matches and Crosslinks as pandas DataFrames**

You might want to convert your crosslink-spectrum-matches and crosslinks to a `pandas.DataFrame` to work with [pandas](https://pandas.pydata.org/) or simply to save them to a file, for example to process them somewhere else. For this purpose the `transform.to_dataframe()` function exists: it takes a list of crosslink-spectrum-matches or crosslinks as input and returns them as a `pandas.DataFrame`. You can read more about the `to_dataframe()` function and its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.to_dataframe.to_dataframe).

In [10]:
df = transform.to_dataframe(csms)
df.head(5)

,Completeness,Alpha Peptide,Alpha Peptide Modifications,Alpha Peptide Crosslink Position,Alpha Proteins,Alpha Proteins Crosslink Positions,Alpha Proteins Peptide Positions,Alpha Score,Alpha Decoy,Beta Peptide,...,Beta Proteins Peptide Positions,Beta Score,Beta Decoy,Crosslink Type,CSM Score,Spectrum File,Scan Nr,Precursor Charge,Retention Time,Ion Mobility
0,full,GQKNSR,(3:[DSS|138.06808]),3,Cas9,779,777,119.825490,False,GQKNSR,...,777,119.825478,False,intra,119.825478,XLpeplib_Beveridge_QEx-HFX_DSS_R1.raw,2257,3,733.18956,0.0
1,full,GQKNSR,(3:[DSS|138.06808]),3,Cas9,779,777,108.852455,False,GSQKDR,...,693,13.909301,True,intra,13.909301,XLpeplib_Beveridge_QEx-HFX_DSS_R1.raw,2448,3,794.66748,0.0
2,full,SDKNR,(3:[DSS|138.06808]),3,Cas9,866,864,114.428277,False,SDKNR,...,864,114.428279,False,intra,114.428277,XLpeplib_Beveridge_QEx-HFX_DSS_R1.raw,2561,3,831.32064,0.0
3,full,DKQSGK,(2:[DSS|138.06808]),2,Cas9,677,676,200.976009,False,DKQSGK,...,676,200.976004,False,intra,200.976004,XLpeplib_Beveridge_QEx-HFX_DSS_R1.raw,2719,3,882.44658,0.0
4,full,DKQSGK,(2:[DSS|138.06808]),2,Cas9,677,676,94.470614,False,HSIKK,...,45,156.884113,False,intra,94.470614,XLpeplib_Beveridge_QEx-HFX_DSS_R1.raw,2792,3,905.68764,0.0


Crosslink-spectrum-matches converted to a `pandas.DataFrame`.

In [11]:
loaded_csms = transform.from_dataframe(df)
_ = transform.summary(loaded_csms)

Reading CSMs...: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 826/826 [00:00<00:00, 2650.47it/s]

Number of CSMs: 826.0
Number of unique CSMs: 826.0
Number of intra CSMs: 803.0
Number of inter CSMs: 23.0
Number of target-target CSMs: 786.0
Number of target-decoy CSMs: 39.0
Number of decoy-decoy CSMs: 1.0
Minimum CSM score: 1.1132827525593785
Maximum CSM score: 452.9861536355926


We can also load back our crosslink-spectrum-matches from the `pandas.DataFrame` using `transfrom.from_dataframe()`. You can read more about the `from_dataframe()` function and its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.from_dataframe.from_dataframe).

In [12]:
df = transform.to_dataframe(xls)
df.head(5)

,Completeness,Alpha Peptide,Alpha Peptide Crosslink Position,Alpha Proteins,Alpha Proteins Crosslink Positions,Alpha Decoy,Beta Peptide,Beta Peptide Crosslink Position,Beta Proteins,Beta Proteins Crosslink Positions,Beta Decoy,Crosslink Type,Crosslink Score
0,full,GQKNSR,3,Cas9,779,False,GQKNSR,3,Cas9,779,False,intra,119.825478
1,full,GQKNSR,3,Cas9,779,True,GSQKDR,4,Cas9,696,True,intra,13.909301
2,full,SDKNR,3,Cas9,866,False,SDKNR,3,Cas9,866,False,intra,114.428277
3,full,DKQSGK,2,Cas9,677,False,DKQSGK,2,Cas9,677,False,intra,200.976004
4,full,DKQSGK,2,Cas9,677,False,HSIKK,4,Cas9,48,False,intra,94.470614


Similarly, we can convert our crosslinks to a `pandas.DataFrame` with `transform.to_dataframe()`...

In [13]:
loaded_xls = transform.from_dataframe(df)
_ = transform.summary(loaded_xls)

Reading crosslinks...: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [00:00<00:00, 7895.46it/s]

Number of crosslinks: 300.0
Number of unique crosslinks by peptide: 300.0
Number of unique crosslinks by protein: 298.0
Number of intra crosslinks: 279.0
Number of inter crosslinks: 21.0
Number of target-target crosslinks: 265.0
Number of target-decoy crosslinks: 0.0
Number of decoy-decoy crosslinks: 35.0
Minimum crosslink score: 1.1132827525593785
Maximum crosslink score: 452.9861536355926


...and load them back with `transform.from_dataframe()`.

## **Crosslink-Spectrum-Matches and Crosslinks in ProForma Notation**

If you want to export your crosslink-spectrum-matches or crosslinks to [ProForma](https://www.psidev.info/proforma) notation, you can do that with `transform.to_proforma()` which you can read more about here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.to_proforma.to_proforma).

Let's go through a few examples:

In [14]:
csm = csms[0]
csm

{'data_type': 'crosslink-spectrum-match',
 'completeness': 'full',
 'alpha_peptide': 'GQKNSR',
 'alpha_modifications': {3: ('DSS', 138.06808)},
 'alpha_peptide_crosslink_position': 3,
 'alpha_proteins': ['Cas9'],
 'alpha_proteins_crosslink_positions': [779],
 'alpha_proteins_peptide_positions': [777],
 'alpha_score': 119.82548987540834,
 'alpha_decoy': False,
 'beta_peptide': 'GQKNSR',
 'beta_modifications': {3: ('DSS', 138.06808)},
 'beta_peptide_crosslink_position': 3,
 'beta_proteins': ['Cas9'],
 'beta_proteins_crosslink_positions': [779],
 'beta_proteins_peptide_positions': [777],
 'beta_score': 119.82547820493929,
 'beta_decoy': False,
 'crosslink_type': 'intra',
 'score': 119.82547820493929,
 'spectrum_file': 'XLpeplib_Beveridge_QEx-HFX_DSS_R1.raw',
 'scan_nr': 2257,
 'charge': 3,
 'retention_time': 733.1895599999999,
 'ion_mobility': 0.0,
 'additional_information': None}

Let's consider the above crosslink-spectrum-match of a homeotypic crosslink where both peptides are `GQKSNR` crosslinked with DSS at the `K3`. 

In [15]:
csm_proforma = transform.to_proforma(csm)

Calling `to_proforma()` on a crosslink-spectrum-match will give us the crosslink-spectrum-match in ProForma notation (as a `str`).

In [16]:
csm_proforma

'GQK[+138.06808]NSR//GQK[+138.06808]NSR/3'

The crosslink modifications are automatically denoted with their corresponding monoisotopic delta masses, for DSS that is `+138.06808` Da.

In [17]:
csms_proforma = transform.to_proforma(csms)

We can also pass a list of crosslink-spectrum-matches to `to_proforma()`...

In [18]:
csms_proforma[:5]

['GQK[+138.06808]NSR//GQK[+138.06808]NSR/3',
 'GQK[+138.06808]NSR//GSQK[+138.06808]DR/3',
 'SDK[+138.06808]NR//SDK[+138.06808]NR/3',
 'DK[+138.06808]QSGK//DK[+138.06808]QSGK/3',
 'DK[+138.06808]QSGK//HSIK[+138.06808]K/3']

...which will return a list of `str` instead, containing the corresponding crosslink-spectrum-matches in ProForma notation.

In [19]:
xl = xls[0]
xl

{'data_type': 'crosslink',
 'completeness': 'full',
 'alpha_peptide': 'GQKNSR',
 'alpha_peptide_crosslink_position': 3,
 'alpha_proteins': ['Cas9'],
 'alpha_proteins_crosslink_positions': [779],
 'alpha_decoy': False,
 'beta_peptide': 'GQKNSR',
 'beta_peptide_crosslink_position': 3,
 'beta_proteins': ['Cas9'],
 'beta_proteins_crosslink_positions': [779],
 'beta_decoy': False,
 'crosslink_type': 'intra',
 'score': 119.82547820493929,
 'additional_information': None}

Alternatively, consider the above crosslink which we get from the previous crosslink-spectrum-match: again both peptides are `GQKNSR` and are crosslinked at the `K3` residue.

In [20]:
xl_proforma = transform.to_proforma(xl)

We can get the ProForma `str` of the crosslink by passing the crosslink as the first argument to `to_proforma()`.

In [21]:
xl_proforma

'GQKNSR//GQKNSR'

Because the crosslink in itself does not contain any information about post-translational-modifications since it is an aggregation of crosslink-spectrum-matches, the returned ProForma string does not contain the crosslinker.

In [22]:
xl_proforma = transform.to_proforma(xl, crosslinker="Xlink:DSS")
xl_proforma

'GQK[Xlink:DSS]NSR//GQK[Xlink:DSS]NSR'

We can explicitly pass the name or mass of the crosslinker via the `crosslinker` parameter to include it in the ProForma string.

In [23]:
csm_proforma = transform.to_proforma(csm, crosslinker="Xlink:DSS")
csm_proforma

'GQK[+138.06808]NSR//GQK[+138.06808]NSR/3'

Passing a crosslinker via the parameter `crosslinker` when there is already a crosslinker specified, as for example in our previous crosslink-spectrum-match, will have no effect.

In [24]:
xls_proforma = transform.to_proforma(xls)

We can also pass a list of crosslinks to `to_proforma()`...

In [25]:
xls_proforma[:5]

['GQKNSR//GQKNSR',
 'GQKNSR//GSQKDR',
 'SDKNR//SDKNR',
 'DKQSGK//DKQSGK',
 'DKQSGK//HSIKK']

...which will return a list of `str` instead, containing the corresponding crosslinks in ProForma notation.